### ISSUE TITLE
Map interaction is blocked when HTML Elements are clicked

### ISSUE THREAD

**Issue by DwieDima** [capacitor-community/capacitor-googlemaps-native#132]
**Describe the bug**
**!! This issue is related to the latest build [v2.0.0-alpha.6](https://github.com/DutchConcepts/capacitor-google-maps/releases/tag/v2.0.0-alpha.6) from the fork of @tafelnl !!**

when the map is initialized, any native element of the map can be interacted with. however, as soon as an HTML element like fab-button, menu, modal etc. is clicked, interaction with the map is blocked.
Furthermore a slight input lag is noticeable after interacting with html.

- on ion-fab buttons this only occurs when you long press on it. Short press works.
- on ion-modal click (click the "+" ion-fab-button to open modal) on ion-header or ion-content will block the interaction. completely. But if you first open the modal, then interact with the map behind and then with ion-modal, the map will work still.
- the same behaviour of ion-modal also goes for the sidemenu.

**To Reproduce**
link to repo:
https://github.com/DwieDima/capacitor-maps-example

**Expected behavior**
interaction should also be possible after clicking on html elements with no input lag.

**Screenshots**
***interaction works with modal, when map is clicked first***
![RPReplay_Final1644398954](https://user-images.githubusercontent.com/26873275/153169715-08332891-f133-4c41-9fcf-9f83276b2669.gif)

***interaction breaks with modal, when modal is clicked first***
![RPReplay_Final1644398979](https://user-images.githubusercontent.com/26873275/153169905-66a1f299-0e07-4371-a580-994761a933e5.gif)

***interaction breaks on sidemenu***
![RPReplay_Final1644399016](https://user-images.githubusercontent.com/26873275/153170303-312e0d31-7773-4a5a-8678-fae7690133a1.gif)

**Additional context**
- IOS

**Comment by tafelnl** [comment]
We also have this some times. We could not reproduce it very reliably, sometimes it starts working again (e.g. after clicking the "my location" button). It seems to be a bit random. But since it's code, it most probably isn't. But we just could not reproduce it the exact same way every time, it always differs a little. What about you?

**Comment by DwieDima** [comment]
This happens to me every time I follow the steps described above.
Clicking on the mylocation button won't fix it on my side. I have to restart the app.

**Comment by DwieDima** [comment]
@tafelnl i managed to get a workaround for this by adding `ontouchstart` event listerner for the div container after map is initialized.

```typescript

      element.ontouchstart = function (ev) {
        CapacitorGoogleMaps.updateMap({
          mapId: result.googleMap.mapId,
        });
      };
```

on the first tap, there is still a input lag, but the interaction works

**Comment by va2ron1** [comment]
@DwieDima Could you try using my [implementation](https://github.com/va2ron1/capacitor-googlemaps-native), because I think this is a problem on how the custom view is implemented. I know that this is not the @tafelnl fork but maybe we can a find a solution there to fix this.

**Comment by DwieDima** [comment]
@va2ron1 wow, it works here as expected and very smoothly! (modal lag is from bad recording)
No issues like before, no input lag, no blocking after clicking on HTML elements.
@tafelnl maybe this would be worth to be reviewed?

![RPReplay_Final1644427261](https://user-images.githubusercontent.com/26873275/153255421-a2613e77-1844-41f8-83c0-643824c0a850.gif)

**Comment by tafelnl** [comment]
Awesome. Definitely going to take a look at it (I am on my phone right now). Could you explain the difference between our and your implementation @va2ron1 ? And has yours any cons? Or is it just better haha

**Comment by va2ron1** [comment]
@tafelnl Honestly I have not fully tested your implementation, apart from a few lines that are fixes for Ionic, I think that both has pro and cons. In my implementation I replaced the bridged view a custom view with the webView and the mapView and I used both views has reference to return the correct view in the hitTest. Your implementation just add the mapView to the webView and that is really a problem.

```
// this is the same as 
self.bridge?.viewController?.view = self.bridge?.viewController?.webView
// So, this is adding the mapView inside the webView
self.bridge?.viewController?.view.addSubview(customMapView.view)
```

**Comment by tafelnl** [comment]
Ahh I think I understand now. So a con of your implementation would be that other plugins/libraries/ capacitor itself, might not have the correct reference for the viewcontroller anymore (since its now the `CustomMapView` is assigned to it instead). I wonder if that could actually break something in practice. Did you encounter scenarios where it did break something?

**Comment by va2ron1** [comment]
I can’t say, I haven’t used a lot of plugins with Capacitor like I used to with Cordova and my current project with this plugin I haven’t found any issue.

**Comment by tafelnl** [comment]
Will try it out. Does this have support for multiple maps being created? Because when a new map is created, it overwrites the viewcontroller right?

**Comment by tafelnl** [comment]
Awesome! I took a quick peek at it already and it looks really good. I'll see if I can test and merge this soon!

**Comment by tafelnl** [comment]
@va2ron1 Would it be possible to have support for multiple map instances at the same time?

Btw, I cleaned some commits and did some rebases. So all your changes are now present on this branch: https://github.com/DutchConcepts/capacitor-google-maps/commits/feat/ios-mapview-behind-webview

**Comment by va2ron1** [comment]
@tafelnl I made a few changes and now support multiple map instances
https://github.com/va2ron1/capacitor-googlemaps-native/tree/development/ios 

**Note**: This assume that you called CapacitorGoogleMaps.initialize per view/page.
**A Better Solution Could Be**: Reset the map views when the view/page change, but I tried a few things and I didn't get any results. Also Capacitor has this [shouldOverrideLoad](https://capacitorjs.com/docs/plugins/ios#override-navigation) and there we can know the current URL and if it change, reset the custom map views, but I get the same URL with/without a Framework.

**Comment by va2ron1** [comment]
@tafelnl Since you [mentioned](https://github.com/capacitor-community/capacitor-googlemaps-native/issues/132#issuecomment-1035216018). I added custom marker icon on Android, doesn't have cache like iOS.
https://github.com/va2ron1/capacitor-googlemaps-native/tree/development/android

**Comment by jjozwiak** [comment]
@va2ron1 @tafelnl I'd like to test this fix out on iOS and it's not clear which repo and branch I should be using. Please advise. Thank you!

**Comment by va2ron1** [comment]
@jjozwiak At some point, everything will be in main repo, is being prepared to be published. For now, is [here](https://github.com/DutchConcepts/capacitor-google-maps/pull/8).

**Comment by jjozwiak** [comment]
@va2ron1 Got it working. This is great. Thanks for your work on this!

**Comment by va2ron1** [comment]
@tafelnl [this branch](https://github.com/capacitor-community/capacitor-googlemaps-native/tree/feat/ios-mapview-behind-webview) doesn't have all the commits from this [pull request](https://github.com/DutchConcepts/capacitor-google-maps/pull/8).

Otherwise, using all the corresponding commits, the examples from [this branch](https://github.com/DutchConcepts/capacitor-google-maps-examples/tree/ios-test-view) didn't work as expected and is because the pixel validation in the [hitTest](https://github.com/va2ron1/capacitor-googlemaps-native/blob/d91adf244bae12bf62a6080702697db3eb347831/ios/Plugin/CustomMapViewController.swift#L15) has the wrong point, need to be changed from _convertedPoint_ to _point_

I would made and submit the change but I don't know to what branch or repo.

**Comment by tafelnl** [comment]
Thanks for your help. Was able to get it to work now :dancing_women: And it works superb :+1:

I updated [the `feat/ios-mapview-behind-webview` branch](https://github.com/capacitor-community/capacitor-googlemaps-native/commits/feat/ios-mapview-behind-webview) with the correct commits. I removed the ability to add a custom background to the scrollview as it is not in the scope of this project right now. (unless it breaks something of course, but as far as I could understand and test, it doesn't).

Also I added a new example in which it is all working [here](https://github.com/DutchConcepts/capacitor-google-maps-examples/tree/ios-test-view-2).

Only thing that's still causing issues, is when a certain div overflows and thus enables scrolling for that div. When that div gets scrolled on top of an already existing mapview instance, the touch gets delegated to the mapview, instead of the div that's in front of it. I made an example for this as well [here](https://github.com/DutchConcepts/capacitor-google-maps-examples/tree/ios-test-view-scroll).

**Comment by va2ron1** [comment]
I found the problem with the way I did it. In simple words the view I'm getting is the view from the web as it is, because of that the size of the view is not same size as the screen and the point I'm getting doesn't have the scroll offset on it and the pixel color take the wrong spot. There're other ways to do this but Capacitor limit us by not wrapping the WKWebView. 

Having all the map views inside the WebView make things more harder to find the better solution.

My solution was to use WKContentView that only contains the HTML Elements and get the right pixel, this view is always the subview at index 0 without the map views but I'm assuming that the api from WKWebView will be always like that in case that change this will break.

The changes are [here](https://github.com/va2ron1/capacitor-googlemaps-native/tree/feat/ios-mapview-behind-webview). I also tested it with all the examples.
